seal.go

```diff
@@ -210,65 +233,126 @@ func SealKeyToTPM(tpm *TPMConnection, key []byte, keyPath string, params *KeyCre
+		if err != nil {
+			return nil, xerrors.Errorf("cannot compute name of signing key for dynamic policy authorization: %w", err)
+		}
+		authKey = goAuthKey.D.Bytes()
+
+		// Create PCR policy counter, if requested.
+		var pcrPolicyCounterPub *tpm2.NVPublic
+		if params.PCRPolicyCounterHandle != tpm2.HandleNull {
+			pcrPolicyCounterPub, err = createPcrPolicyCounter(tpm.TPMContext, params.PCRPolicyCounterHandle, authKeyName, session)
+			switch {
+			case tpm2.IsTPMError(err, tpm2.ErrorNVDefined, tpm2.CommandNVDefineSpace):
+				return nil, TPMResourceExistsError{params.PCRPolicyCounterHandle}
+			case isAuthFailError(err, tpm2.CommandNVDefineSpace, 1):
+				return nil, AuthFailError{tpm2.HandleOwner}
+			case err != nil:
+				return nil, xerrors.Errorf("cannot create new dynamic authorization policy counter: %w", err)
+			}
+			defer func() {
+				if succeeded {
+					return
+				}
+				index, err := tpm2.CreateNVIndexResourceContextFromPublic(pcrPolicyCounterPub)
+				if err != nil {
+					return
+				}
+				tpm.NVUndefineSpace(tpm.OwnerHandleContext(), index, session)
+			}()
+		}
 
-	template := makeSealedKeyTemplate()
+		metadataVersion = currentMetadataVersion
+		template = makeSealedKeyTemplate()
 
-	// Compute the static policy - this never changes for the lifetime of this key file
-	staticPolicyData, authPolicy, err := computeStaticPolicy(template.NameAlg, &staticPolicyComputeParams{
-		key:                 authPublicKey,
-		pcrPolicyCounterPub: pcrPolicyCounterPub})
-	if err != nil {
-		return nil, xerrors.Errorf("cannot compute static authorization policy: %w", err)
-	}
+		// Compute the static policy - this never changes for the lifetime of this key file
+		var authPolicy tpm2.Digest
+		staticPolicyData, authPolicy, err = computeStaticPolicy(template.NameAlg, &staticPolicyComputeParams{
+			key:                 authPublicKey,
+			pcrPolicyCounterPub: pcrPolicyCounterPub})
+		if err != nil {
+			return nil, xerrors.Errorf("cannot compute static authorization policy: %w", err)
+		}
 
-	// Define the template for the sealed key object, using the computed policy digest
-	template.AuthPolicy = authPolicy
+		// Define the template for the sealed key object, using the computed policy digest
+		template.AuthPolicy = authPolicy
+
+		// Create a dynamic authorization policy
+		pcrProfile := params.PCRProfile
+		if pcrProfile == nil {
+			pcrProfile = &PCRProtectionProfile{}
+		}
+		dynamicPolicyData, err = computeSealedKeyDynamicAuthPolicy(tpm.TPMContext, currentMetadataVersion, template.NameAlg,
+			authPublicKey.NameAlg, goAuthKey, pcrPolicyCounterPub, nil, pcrProfile, session)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot compute dynamic authorization policy: %w", err)
+		}
+
+		// Increment the PCR policy counter for the first time.
+		if pcrPolicyCounterPub != nil {
+			if err := incrementPcrPolicyCounter(tpm.TPMContext, currentMetadataVersion, pcrPolicyCounterPub, nil, goAuthKey, authPublicKey,
```

Comment:
we are incrementing the counter before the tpm.Create now, not sure it makes a big difference but wanted to point it out. Anyway in the refactoring to a single to wrap many keys it can probably be changed back more easily

Reply:
It doesn't matter where it is because the counter gets undefined on failure, but it is easier to move it back to the end of the function in the case where related keys are created in one call, so I'll do that.